kitsune/users/api.py

```diff
@@ -54,26 +53,48 @@ def usernames(request):
     if not request.user.is_authenticated:
         return []
 
-    profiles = Profile.objects.filter(Q(name__istartswith=pre)).values_list("user_id", flat=True)
+    profile_ids = list(
+        Profile.objects.filter(Q(name__istartswith=pre)).values_list("user_id", flat=True)[:10]
+    )
     users = (
-        User.objects.filter(Q(username__istartswith=pre) | Q(id__in=profiles))
-        .extra(select={"length": "Length(username)"})
-        .order_by("length")
+        User.objects.filter(Q(username__istartswith=pre) | Q(id__in=profile_ids))
+        .filter(is_active=True)
         .select_related("profile")
     )
 
-    if not waffle.switch_is_active("users-dont-limit-by-login"):
-        last_login = datetime.now() - timedelta(weeks=12)
-        users = users.filter(last_login__gte=last_login)
+    autocomplete_list = []
+    exact_match_in_list = False
+
+    for u in users[:10]:
+        if u.username.lower() == pre.lower():
+            exact_match_in_list = True
+        autocomplete_list.append(
+            {
+                "username": u.username,
+                "display_name": display_name_or_none(u),
+                "avatar": profile_avatar(u, 24),
+            }
+        )
 
-    return [
-        {
-            "username": u.username,
-            "display_name": display_name_or_none(u),
-            "avatar": profile_avatar(u, 24),
-        }
-        for u in users[:10]
-    ]
+    if not exact_match_in_list:
```

Comment:
Using `icontains` should be enough. No need for this extra step at all

Reply:
I presume `icontains` will work the same as `istartswith` in this regard (but I'll test it). mysql will return as soon as it finds any 10 rows which match the query, so if the exact match is "slower" to access than any other 10 rows it won't end up in the result, which breaks the functionality on the frontend. This is why the previous code was sorting by length, so the exact match would always appear first (but massively at the expense of performance).

I'll test with `icontains` and if it doesn't always return the exact match add a comment here clarifying why it's necessary.